br/pkg/restore/split/client.go

```diff
@@ -555,39 +572,68 @@ func (c *pdClient) hasHealthyRegion(ctx context.Context, regionID uint64) (bool,
 	return len(regionInfo.PendingPeers) == 0, nil
 }
 
-func (c *pdClient) SplitWaitAndScatter(
-	ctx context.Context, region *RegionInfo, keys [][]byte,
-) (*RegionInfo, []*RegionInfo, error) {
+func (c *pdClient) SplitWaitAndScatter(ctx context.Context, region *RegionInfo, keys [][]byte) ([]*RegionInfo, error) {
 	failpoint.Inject("failToSplit", func(_ failpoint.Value) {
-		failpoint.Return(nil, nil, errors.New("retryable error"))
+		failpoint.Return(nil, errors.New("retryable error"))
 	})
 	if len(keys) == 0 {
-		return region, []*RegionInfo{region}, nil
+		return []*RegionInfo{region}, nil
 	}
 
-	origin, newRegions, err := c.batchSplitRegionsWithOrigin(ctx, region, keys)
-	if err != nil {
-		return nil, nil, errors.Trace(err)
-	}
-	err = c.waitRegionsSplit(ctx, newRegions)
-	if err != nil {
-		brlog.FromContext(ctx).Warn(
-			"wait regions split failed, will continue anyway",
-			zap.Error(err),
-		)
-	}
-	if err = ctx.Err(); err != nil {
-		return nil, nil, errors.Trace(err)
-	}
+	var (
+		start, end = 0, 0
+		batchSize  = 0
+		newRegions = make([]*RegionInfo, 0, len(keys))
+	)
 
-	err = c.scatterRegions(ctx, newRegions)
-	if err != nil {
-		brlog.FromContext(ctx).Warn(
-			"scatter regions failed, will continue anyway",
-			zap.Error(err),
-		)
+	for end <= len(keys) {
+		if end == len(keys) ||
+			batchSize+len(keys[end]) > maxBatchSplitSize ||
+			end-start >= c.splitBatchKeyCnt {
+			// split, wait and scatter for this batch
+			originRegion, newRegionsOfBatch, err := c.batchSplitRegionsWithOrigin(ctx, region, keys[start:end])
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			err = c.waitRegionsSplit(ctx, newRegionsOfBatch)
+			if err != nil {
+				brlog.FromContext(ctx).Warn(
+					"wait regions split failed, will continue anyway",
+					zap.Error(err),
+				)
+			}
+			if err = ctx.Err(); err != nil {
+				return nil, errors.Trace(err)
+			}
+			err = c.scatterRegions(ctx, newRegionsOfBatch)
+			if err != nil {
+				brlog.FromContext(ctx).Warn(
+					"scatter regions failed, will continue anyway",
+					zap.Error(err),
+				)
+			}
+
+			// the region with the max start key is the region need to be further split,
+			// depending on the origin region is the first region or last region, we need to
+			// compare the origin region and the last one of new regions.
+			lastNewRegion := newRegionsOfBatch[len(newRegionsOfBatch)-1]
```

Comment:
it will panic if the `newRegionsOfBatch` is nil when `batchSplitRegionsWithOrigin` returns error.

Reply:
we have checked the error in line 595, that error check will return